git/keygen.go

```diff
@@ -54,31 +56,101 @@ func MakeKeyPair() (*KeyPair, error) {
 	return &KeyPair{privStr, pubStr}, nil
 }
 
-// PrivKeyPath returns the full path for the location of the user's private key file.
-func PrivKeyPath(user string) string {
+// PrivKeyPath returns a map with the full path for all the currently available private key files indexed by the server alias for each key.
+func PrivKeyPath() map[string]string {
 	configpath, err := config.Path(false)
 	if err != nil {
 		log.Write("Error getting user's config path. Can't load key file.")
 		log.Write(err.Error())
-		return ""
+		return nil
+	}
+	servers := config.Read().Servers
+	keys := make(map[string]string)
+	for srvalias := range servers {
+		keyfilepath := filepath.Join(configpath, fmt.Sprintf("%s.key", srvalias))
+		if pathExists(keyfilepath) {
+			keys[srvalias] = keyfilepath
+		}
 	}
-	return filepath.Join(configpath, fmt.Sprintf("%s.key", user))
+	return keys
 }
 
-// HostKeyPath returns the full path for the location of the gin host key file.
-func HostKeyPath() string {
+// GetHostKey takes a git server configuration, queries the server via SSH, and returns the public key of the host (in the format required for the known_hosts file) and the key fingerprint.
+func GetHostKey(gitconf config.GitCfg) (hostkeystr, fingerprint string, err error) {
+	// HostKeyCallback constructs the keystring
+	keycb := func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+		hostkeystr = fmt.Sprintf("%s", gitconf.Host)
+		if gitconf.Port != 22 {
+			// Only specify if non-standard port
+			hostkeystr = fmt.Sprintf("%s:%d", hostkeystr, gitconf.Port)
+		}
+		ip := remote.String()
+		if strings.HasSuffix(ip, ":22") {
+			// Only specify if non-standard port
+			ip = strings.TrimSuffix(ip, ":22")
+		}
+		hostkeystr = fmt.Sprintf("%s,%s %s", hostkeystr, ip, string(ssh.MarshalAuthorizedKey(key)))
+		hostkeystr = strings.TrimSpace(hostkeystr)
+		fingerprint = ssh.FingerprintSHA256(key)
+		return nil
+	}
+
+	sshcon := ssh.ClientConfig{
+		User:            gitconf.User,
+		HostKeyCallback: keycb,
+	}
+	_, derr := ssh.Dial("tcp", fmt.Sprintf("%s:%d", gitconf.Host, gitconf.Port), &sshcon)
+	if derr != nil && !strings.Contains(derr.Error(), "unable to authenticate") {
+		// Other errors (auth error in particular) should be ignored
+		err = fmt.Errorf("connection test failed: %s", derr)
+	}
+	return
+}
+
+// hostkeypath returns the full path for the location of the gin host key file.
+func hostkeypath() string {
 	configpath, err := config.Path(false)
 	if err != nil {
 		log.Write("Error getting user's config path. Can't create host key file.")
 		log.Write(err.Error())
 		return ""
 	}
-	return filepath.Join(configpath, "ginhostkey")
+	filename := "known_hosts"
+	return filepath.Join(configpath, filename)
+}
+
+// WriteKnownHosts creates a known_hosts file in the config directory with all configured host keys.
+func WriteKnownHosts() error {
+	conf := config.Read()
+	hostkeyfile := hostkeypath()
```

Comment:
Check if `hostkeypath()` threw an error before handing the `hostkeyfile` string to `os.OpenFile`.